osgpkitools/incommon_request.py

```diff
@@ -0,0 +1,453 @@
+    """This function submits an enrollment request for a certificate
+       If successful returns a self-enrollment certificate Id = sslId
+    """
+    # Build and update headers for the restclient. Headers will be reused for all requests 
+    headers = buildHeaders(config)
+
+    response = None
+    response_data = None
+
+    cert_type = config['igtfservercert']
+    
+    if sans:
+        cert_type = config['igtfmultidomain']
+    
+    payload = dict(
+        csr=cert_csr,
+        orgId=config['department'],
+        certType=cert_type,
+        numberServers=0,
+        serverType=config['servertype'],
+        term=config['term'],
+        comments="Certificate request for " + hostname
+    )
+   
+    if sans:
+        payload.update(subjAltNames=sans)
+    
+    try:
+        response = restclient.post_request(config['enrollurl'], headers, payload)
+        
+        if response.status == 200:
+            response_text = response.read()
+            logger.debug('response text: ' + str(response_text))
+            response_data = json.loads(response_text)
+            response_data = response_data['sslId']
+    except httplib.HTTPException as exc:
+        raise
+    
+    return response_data
+    
+def retrieve_cert(config, sslcontext, sslId):
+    """This function retrieves a certificate given a self-enrollment certificate Id = sslId
+    """
+    
+    # Build and update headers for the restclient. Headers will be reused for all requests
+    headers = buildHeaders(config)
+
+    response = None
+    response_data = None    
+
+    retry_count = MAX_RETRY_RETRIEVAL
+    retrieve_url = config['retrieveurl'] + str(sslId) + config['certx509co']
+    
+    for _ in range(retry_count):
+        try:
+            # If the HTTPSConnection is reused 
+            restclient = InCommonApiClient(config['apiurl'], sslcontext)
+            response = restclient.get_request(retrieve_url, headers)
+            # InCommon API responds with 400 Bad Request when the certificate is still being procesed
+            # "code": 0, "description": "Being processed by Sectigo"
+            # It triggers the BadStatusLine exception avoiding to reuse the HTTPSConnection
+            response_text = response.read()
+            logger.debug('response text: ' + str(response_text))
+            if response.status == 200:
+                response_data = response_text
+                restclient.closeConnection()
+                break
+        except httplib.BadStatusLine as exc:
+            # BadStatusLine is raised as the server responded with a HTTP status code that we don't understand.
+            pass
+        except httplib.HTTPException as exc:
+            raise
+        charlimit_textwrap('    Waiting for %s seconds before retrying certificate retrieval' % WAIT_RETRIEVAL )
+        # Closing the connection before sleeping
+        restclient.closeConnection()
+        time.sleep(WAIT_RETRIEVAL)
```

Comment:
Would an exponential backoff be more appropriate here?

Reply:
I don't think so.  We haven't observed it slowing down when there are multiple requests.  I haven't seen it have to wait more than one extra 5 second delay beyond the 30 seconds.